### PR TITLE
[Fix] fix fast scnn

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Train with current fast scnn config will get low performance, we may fix the config and code.

## Modification

- Set the `loss_weight` of decode head from **0.4 to 1**, then the training result is correct.
- For improving the performance, we match our network code with [paddleseg](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.1/paddleseg/models/fast_scnn.py). The key modification is that we change `conv 1x1` to `conv 3x3 with padding 1`.
- Add a param `dw_act_cfg` at the backbone `__init__` function for customizing `depthwise ConvModule`.

Experiments:

|iters|mIoU|
|---|---|
|160k|70.96|

### CODE REVIEW

File: mmseg/models/backbones/fast_scnn.py [hunk 1]
```diff
@@ -166,14 +174,15 @@ def _make_layer(self,
...
+                    norm_cfg=self.norm_cfg,
+                    act_cfg=dict(type='ReLU')))
         return nn.Sequential(*layers)
 
     def forward(self, x):
         x = self.bottleneck1(x)
         x = self.bottleneck2(x)
         x = self.bottleneck3(x)
-        x = torch.cat([x, *self.ppm(x)], dim=1)
+        x = torch.cat([x, *self.ppm(x)[::-1]], dim=1)
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:185]: `[::-1]` is not necessary if we train from scratch.

File: configs/_base_/datasets/cityscapes.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
         img_scale=(2048, 1024),
         # img_ratios=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75],
         flip=False,
+        # flip=True,
```
Comment (Junjun2016) [inline configs/_base_/datasets/cityscapes.py:26]: Delete `# flip=True`

File: configs/fastscnn/fast_scnn_4x8_160k_lr0.12_cityscapes.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+_base_ = [
+    '../_base_/models/fast_scnn.py', '../_base_/datasets/cityscapes.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
+]
+
+# Re-config the data sampler.
+data = dict(samples_per_gpu=4, workers_per_gpu=4)
```
Comment (Junjun2016) [inline configs/fastscnn/fast_scnn_4x8_160k_lr0.12_cityscapes.py:7]: If we use 8 GPUs and 4 batch size per GPU, please rename config name `4x8` to `8x4`

File: configs/fastscnn/fast_scnn_4x8_320k_lr0.12_cityscapes.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+_base_ = [
+    '../_base_/models/fast_scnn.py', '../_base_/datasets/cityscapes.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
+]
+
+# Re-config the data sampler.
+data = dict(samples_per_gpu=4, workers_per_gpu=4)
```
Comment (Junjun2016) [inline configs/fastscnn/fast_scnn_4x8_320k_lr0.12_cityscapes.py:7]: If we use 8 GPUs and 4 batch size per GPU, please rename config name 4x8 to 8x4

File: mmseg/models/backbones/fast_scnn.py [hunk 5]
```diff
@@ -45,23 +45,28 @@ def __init__(self,
...
+
         self.dsconv1 = DepthwiseSeparableConvModule(
             dw_channels1,
             dw_channels2,
             kernel_size=3,
             stride=2,
             padding=1,
-            norm_cfg=self.norm_cfg)
+            norm_cfg=self.norm_cfg,
+            dw_act_cfg=None)
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:60]: We can have an ablation study for `dw_act_cfg`.

File: mmseg/models/backbones/fast_scnn.py [hunk 6]
```diff
@@ -45,23 +45,28 @@ def __init__(self,
...
+
         self.dsconv2 = DepthwiseSeparableConvModule(
             dw_channels2,
             out_channels,
             kernel_size=3,
             stride=2,
             padding=1,
-            norm_cfg=self.norm_cfg)
+            norm_cfg=self.norm_cfg,
+            dw_act_cfg=None)
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:69]: We can have an ablation study for dw_act_cfg.

File: mmseg/models/backbones/fast_scnn.py [hunk 7]
```diff
@@ -136,11 +141,13 @@ def __init__(self,
             conv_cfg=self.conv_cfg,
             norm_cfg=self.norm_cfg,
             act_cfg=self.act_cfg,
-            align_corners=align_corners)
+            align_corners=True)
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:144]: `align_corners=True` -->  `align_corners=align_corners`
We can also have an ablation study for `align_corners`.

**Review by Junjun2016 — COMMENTED** [review]
Have an ablation study for `DepthwiseSeparableConvModule`, one with `activation` after depthwise convolution and one without.

**Review by Junjun2016 — COMMENTED** [review]
What are the results of our original model with more iteration (160k and 320k).

**Comment by xiexinch** [comment]
> What are the results of our original model with more iteration (160k and 320k).

Training results of master branch model:

|iters|mIoU|
|---|---|
|80k|66.96|
|160k|69.69|
|320k|71.17|

**Comment by xvjiarui** [comment]
We may describe what we changed in the PR message. 
And update the link of the models.

File: configs/fastscnn/fast_scnn_8x4_160k_lr0.12_cityscapes.py [hunk 8]
```diff
@@ -1,10 +1,10 @@
 _base_ = [
```
Comment (Junjun2016) [inline configs/fastscnn/fast_scnn_8x4_160k_lr0.12_cityscapes.py:1]: fast_scnn_8x4_160k_lr0.12_cityscapes.py --> fast_scnn_lr0.12_8x4_160k_cityscapes.py

File: mmseg/models/utils/inverted_residual.py [hunk 9]
```diff
@@ -37,7 +37,8 @@ def __init__(self,
                  conv_cfg=None,
                  norm_cfg=dict(type='BN'),
                  act_cfg=dict(type='ReLU6'),
-                 with_cp=False):
+                 with_cp=False,
+                 **kwards):
```
Comment (Junjun2016) [inline mmseg/models/utils/inverted_residual.py:41]:
```suggestion
                 **kwargs):
```

File: mmseg/models/utils/inverted_residual.py [hunk 10]
```diff
@@ -55,7 +56,8 @@ def __init__(self,
                     kernel_size=1,
                     conv_cfg=conv_cfg,
                     norm_cfg=norm_cfg,
-                    act_cfg=act_cfg))
+                    act_cfg=act_cfg,
+                    **kwards))
```
Comment (Junjun2016) [inline mmseg/models/utils/inverted_residual.py:60]:
```suggestion
                    **kwargs))
```

File: mmseg/models/utils/inverted_residual.py [hunk 11]
```diff
@@ -67,14 +69,16 @@ def __init__(self,
                 groups=hidden_dim,
                 conv_cfg=conv_cfg,
                 norm_cfg=norm_cfg,
-                act_cfg=act_cfg),
+                act_cfg=act_cfg,
+                **kwards),
```
Comment (Junjun2016) [inline mmseg/models/utils/inverted_residual.py:73]:
```suggestion
                **kwargs),
```

File: mmseg/models/utils/inverted_residual.py [hunk 12]
```diff
@@ -67,14 +69,16 @@ def __init__(self,
...
             ConvModule(
                 in_channels=hidden_dim,
                 out_channels=out_channels,
                 kernel_size=1,
                 conv_cfg=conv_cfg,
                 norm_cfg=norm_cfg,
-                act_cfg=None)
+                act_cfg=None,
+                **kwards)
```
Comment (Junjun2016) [inline mmseg/models/utils/inverted_residual.py:81]:
```suggestion
                **kwargs)
```

File: mmseg/models/decode_heads/psp_head.py [hunk 13]
```diff
@@ -22,7 +22,7 @@ class PPM(nn.ModuleList):
     """
 
     def __init__(self, pool_scales, in_channels, channels, conv_cfg, norm_cfg,
-                 act_cfg, align_corners):
+                 act_cfg, align_corners, **kwards):
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/psp_head.py:25]:
```suggestion
                 act_cfg, align_corners, **kwargs):
```

File: mmseg/models/decode_heads/psp_head.py [hunk 14]
```diff
@@ -41,7 +41,8 @@ def __init__(self, pool_scales, in_channels, channels, conv_cfg, norm_cfg,
                         1,
                         conv_cfg=self.conv_cfg,
                         norm_cfg=self.norm_cfg,
-                        act_cfg=self.act_cfg)))
+                        act_cfg=self.act_cfg,
+                        **kwards)))
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/psp_head.py:45]:
```suggestion
                        **kwargs)))
```

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 15]
```diff
@@ -33,19 +33,23 @@ def __init__(self, **kwargs):
...
+
         for i in range(1, self.num_convs):
             self.convs[i] = DepthwiseSeparableConvModule(
                 self.channels,
                 self.channels,
                 kernel_size=self.kernel_size,
                 padding=self.kernel_size // 2,
-                norm_cfg=self.norm_cfg)
+                norm_cfg=self.norm_cfg,
+                dw_act_cfg=None)
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sep_fcn_head.py:46]:
```suggestion
                dw_act_cfg=dw_act_cfg)
```

File: mmseg/models/decode_heads/sep_fcn_head.py [hunk 16]
```diff
@@ -33,19 +33,23 @@ def __init__(self, **kwargs):
...
 
         if self.concat_input:
             self.conv_cat = DepthwiseSeparableConvModule(
                 self.in_channels + self.channels,
                 self.channels,
                 kernel_size=self.kernel_size,
                 padding=self.kernel_size // 2,
-                norm_cfg=self.norm_cfg)
+                norm_cfg=self.norm_cfg,
+                dw_act_cfg=None)
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sep_fcn_head.py:55]:
```suggestion
                dw_act_cfg=dw_act_cfg)
```

File: mmseg/models/backbones/fast_scnn.py [hunk 17]
```diff
@@ -23,6 +23,8 @@ class LearningToDownsample(nn.Module):
             dict(type='BN')
         act_cfg (dict): Config of activation layers. Default:
             dict(type='ReLU')
+        dw_act_cfg (dict):Activation config of depthwise ConvModule. If it is
+            'default', it will be the same as `act_cfg`. Default: None.
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:27]: In `DepthwiseSeparableConvModule`, if it is 'default', it will be the same as `act_cfg`.

File: mmseg/models/backbones/fast_scnn.py [hunk 18]
```diff
@@ -290,6 +309,8 @@ class FastSCNN(BaseModule):
             dict(type='ReLU')
         align_corners (bool): align_corners argument of F.interpolate.
             Default: False
+        dw_act_cfg (dict): Activation config of depthwise ConvModule. If it is
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:312]: Same

File: mmseg/models/backbones/fast_scnn.py [hunk 19]
```diff
@@ -364,7 +387,7 @@ def __init__(self,
             fusion_out_channels,
             conv_cfg=self.conv_cfg,
             norm_cfg=self.norm_cfg,
-            act_cfg=self.act_cfg,
+            dwconv_act_cfg=self.act_cfg,
```
Comment (Junjun2016) [inline mmseg/models/backbones/fast_scnn.py:390]: act_cfg?

Comment (xiexinch) [reply mmseg/models/backbones/fast_scnn.py:390]: `dwconv_act_cfg` is the `act_cfg` in original version

File: mmseg/models/backbones/fast_scnn.py (hunk 18 above)
Comment (xiexinch) [reply mmseg/models/backbones/fast_scnn.py:312]: `'default'` is a string value at `DepthwiseSeparableConvModule`. In `LearningToDownsample` module, I want to set it to `None`.
